plugins/ingest-useragent/src/main/java/org/elasticsearch/ingest/useragent/UseragentParser.java

```diff
@@ -0,0 +1,276 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.ingest.useragent;
+
+import org.elasticsearch.ElasticsearchParseException;
+import org.elasticsearch.common.cache.Cache;
+import org.elasticsearch.common.cache.CacheBuilder;
+import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class UseragentParser {
+
+    private Cache<String, Details> cache = null;
+    private List<UseragentSubpattern> uaPatterns;
+    private List<UseragentSubpattern> osPatterns;
+    private List<UseragentSubpattern> devicePatterns;
+    
+    public UseragentParser(InputStream regexStream, int cacheSize) {
+        try {
+            init(regexStream);
+        } catch (IOException e) {
+            throw new ElasticsearchParseException("error parsing regular expression file", e);
+        }
+        
+        if (cacheSize > 0) {
+            cache = CacheBuilder.<String, Details>builder().setMaximumWeight(cacheSize).build();
+        }
+    }
+    
+    private void init(InputStream regexStream) throws IOException {
+        XContentParser yamlParser = XContentFactory.xContent(XContentType.YAML).createParser(regexStream);
```

Comment:
Can we merge the this method into the constructor? If so then we can make all the fields final.